### PR TITLE
Fix #388 hrr test case failed in sha384 with openssl peer

### LINKED ISSUE #388: ssl-opt.sh failes with HRR test cases if we change the server peer to OpenSSL s_server
There are two exist test cases for HRR in ssl-opt.sh, with both peers ssl_client and ssl_server.  Both of them can pass the test cases even we change the cipher suite. 
But when we change the server peer from ssl_server to openssl s_server, it can pass with cipher suite TLS_AES_128_GCM_SHA256, but failes with the cipher suite TLS_AES_256_GCM_SHA384.

### System information:
Latest code from branch: tls13-prototype

### Steps to reproduce
Add the following commands into the ssl-opt.sh.

requires_config_enabled MBEDTLS_SSL_PROTO_TLS1_3_EXPERIMENTAL
requires_config_enabled MBEDTLS_DEBUG_C
requires_config_enabled MBEDTLS_SSL_CLI_C
run_test    "TLS 1.3 O-M, TLS1-3-AES-256-GCM-SHA384 with ECDHE-ECDSA, SRV auth, HRR enforcing group" \
            "$O_SRV -msg -tls1_3 -ciphersuites TLS_AES_256_GCM_SHA384 -groups P-256 -num_tickets 0 -no_resume_ephemeral -no_cache" \
            "$P_CLI debug_level=4 force_version=tls1_3 server_name=localhost force_ciphersuite=TLS1-3-AES-256-GCM-SHA384 tls13_kex_modes=ephemeral" \
            0 \
            -c "received HelloRetryRequest message"       \
            -c "Protocol is TLSv1.3"                      \
            -c "Ciphersuite is TLS1-3-AES-256-GCM-SHA384" \
            -c "Verifying peer X.509 certificate... ok"

### PR DESCRIPTION
**Pull request by xkqian**

### CODE REVIEW

File: library/ssl_tls.c [hunk 1]
```diff
@@ -634,40 +634,59 @@ int mbedtls_ssl_reset_transcript_for_hrr( mbedtls_ssl_context *ssl )
...
 
     MBEDTLS_SSL_DEBUG_MSG( 3, ( "Reset SSL session for HRR" ) );
 
-#if defined(MBEDTLS_SHA256_C)
-    ret = ssl_hash_transcript_core( ssl, MBEDTLS_MD_SHA256,
-                                    hash_transcript,
-                                    sizeof( hash_transcript ),
-                                    &hash_olen );
-    if( ret != 0 )
+    if( ciphersuite_info->mac == MBEDTLS_MD_SHA256 )
```
Comment (yuhaoth) [inline library/ssl_tls.c:643]: I think it is not necessary . 
SHA256 will update `handshake->fin_sha256`. And SHA384 will update `handshake->fin_sha512`

File: tests/ssl-opt.sh [hunk 2]
```diff
@@ -2525,6 +2525,20 @@ run_test    "TLS 1.3, TLS1-3-AES-128-CCM-SHA256 with ECDHE-ECDSA, SRV auth, HRR
             -c "Ciphersuite is TLS1-3-AES-128-CCM-SHA256" \
             -c "Verifying peer X.509 certificate... ok"
 
+# configure client to initially sent incorrect group, which will be corrected with HRR from the server
```
Comment (yuhaoth) [inline tests/ssl-opt.sh:2528]: Add test with AES-128-GCM-SHA256 to make sure SHA256 will work correctly

File: library/ssl_tls.c [hunk 3]
```diff
@@ -634,40 +634,59 @@ int mbedtls_ssl_reset_transcript_for_hrr( mbedtls_ssl_context *ssl )
...
-                                    sizeof( hash_transcript ),
-                                    &hash_olen );
-    if( ret != 0 )
-    {
-        MBEDTLS_SSL_DEBUG_RET( 4, "ssl_hash_transcript_core", ret );
-        return( ret );
     }
-    MBEDTLS_SSL_DEBUG_BUF( 4, "Truncated SHA-384 handshake transcript",
-                           hash_transcript, hash_olen );
+    else if( ciphersuite_info->mac == MBEDTLS_MD_SHA384 )
```
Comment (yuhaoth) [inline library/ssl_tls.c:667]: ditto

**Review by yuhaoth — CHANGES_REQUESTED** [review]

File: library/ssl_tls.c (hunk 1 above)
Comment (xkqian) [reply library/ssl_tls.c:643]: It's necessary that we reset sha256 or sha384 seperately but not both.

File: library/ssl_tls.c (hunk 3 above)
Comment (xkqian) [reply library/ssl_tls.c:667]: It's necessary that we reset sha256 or sha384 seperately based on the cipher suite we select but not both.

File: tests/ssl-opt.sh [hunk 4]
```diff
@@ -2525,6 +2525,34 @@ run_test    "TLS 1.3, TLS1-3-AES-128-CCM-SHA256 with ECDHE-ECDSA, SRV auth, HRR
...
+requires_config_enabled MBEDTLS_SSL_PROTO_TLS1_3_EXPERIMENTAL
+requires_config_enabled MBEDTLS_DEBUG_C
+requires_config_enabled MBEDTLS_SSL_CLI_C
+run_test    "TLS 1.3 O-M, TLS1-3-AES-256-GCM-SHA384 with ECDHE-ECDSA, SRV auth, HRR enforcing group" \
```
Comment (yuhaoth) [inline tests/ssl-opt.sh]:
```suggestion
run_test    "TLS 1.3 m->O, TLS1-3-AES-256-GCM-SHA384 with ECDHE-ECDSA, SRV auth, HRR enforcing group" \
```

File: tests/ssl-opt.sh [hunk 5]
```diff
@@ -2525,6 +2525,34 @@ run_test    "TLS 1.3, TLS1-3-AES-128-CCM-SHA256 with ECDHE-ECDSA, SRV auth, HRR
...
+            -c "Protocol is TLSv1.3"                      \
+            -c "Ciphersuite is TLS1-3-AES-256-GCM-SHA384" \
+            -c "Verifying peer X.509 certificate... ok"   \
+            -c "HTTP/1.0 200 ok"
+
+# configure client to initially sent incorrect group, which will be corrected with HRR from the server
+requires_config_enabled MBEDTLS_SSL_PROTO_TLS1_3_EXPERIMENTAL
+requires_config_enabled MBEDTLS_DEBUG_C
+requires_config_enabled MBEDTLS_SSL_CLI_C
+run_test    "TLS 1.3 O-M, TLS1-3-AES-128-GCM-SHA256 with ECDHE-ECDSA, SRV auth, HRR enforcing group" \
```
Comment (yuhaoth) [inline tests/ssl-opt.sh]:
```suggestion
run_test    "TLS 1.3 m->O, TLS1-3-AES-128-GCM-SHA256 with ECDHE-ECDSA, SRV auth, HRR enforcing group" \
```

**Review by yuhaoth — CHANGES_REQUESTED** [review]

**Review by yuhaoth — APPROVED** [review]

**Comment by yuhaoth** [comment]
LGTM

**Pull request merged**